broadinstitute/single_cell_portal_core
# Scalable expression overlays for pathway diagrams (SCP-5710)

This enables pathway visualizations for arbitrarily many genes.  That means more causal insight for scientists.

### Background
[Previously](https://github.com/broadinstitute/single_cell_portal_core/pull/2104), pathway diagram expression overlays were novel, useful, and limited.  They give researchers dataset-specific expression summaries for genes in a biochemical pathway, e.g. a gene regulatory network for human lactation, an immune signaling pathway, etc.

### Problem
Users could only view expression for up to 50 genes in a pathway.  That derived from a simple way to prevent overloading the SCP app server.  However, roughly 30% of pathways have > 50 genes, and 15% of pathways have over 100 genes.  So a basic implementation of the 50-gene maximum really limited the causal mechanistic insight available to researchers.

### Solution
Now, expression data is fetched for _all_ genes in a pathway in _serial batches_ of up to 50 genes at a time.  For example, a pathway with 300 unique genes will get expression metrics loaded 50 genes at a time, one batch starting only after the previous batch completes, until all genes in the pathway are colored in.  

This greatly mitigates the risk of exhausting server CPU, RAM, and threads.  It does that by splitting any oversized queries into multiple smaller queries, and stream-processing them one at a time.

The updates here also pull in upstream improvements in Ideogram.js, to better account for pathway genes that interact as a member of a _gene group_ (i.e., a gene family or complex), and to use updated pathway data.

### Video
Here's how this robustification looks through an engineering lens, and how it yields richer results for big pathways.

https://github.com/user-attachments/assets/eb871da1-c89b-4085-9794-d194140e5a8c

### Next steps
* Apply this streaming-batch-query robustness technique to dot plots / heatmaps, to eliminate _from a user's perspective_ the max limit of number of genes that can be searched, while essentially preserving this limit _from the server's perspective_.
* Diagram nodes indicate they're clickable due to Pvjs, but clicking them does nothing. Ideally the cursor would be changed, or clicking would e.g. show expression metrics, gene information, etc.
* The framing UX seems viable, but a fuller integration into the many-gene display (e.g. 3+, with dot plots as default enabled tab) would be better.

### Test
A new automated test confirms batching behavior.  To manually verify:
1.  `yarn install --force`
2.  Start server without Vite, or clear your frontend service worker cache via Chrome DevTools
3. Go to "Cellular and transcriptional diversity over the course of human lactation"
4. Search CSN2
5. Open DevTools, go to Network panel
6. In related genes ideogram, hover over GSK3B
7. Click "IL8 signaling"
8. In Network panel, confirm several requests to Morpheus get sent, one after another, never overlapping in time.  See 0:07 - 0:20 in video above for reference.  Your local requests will take longer at least the first time, as they aren't even cached in Rails.  Rails-uncached requests can take > 20 seconds each in a local SCP instance.  If you see _no_ Morpheus requests in the Network panel, ensure your Vite server is not running or that you've deleted any frontend service worker cache entries for "morpheus" in the "scp-..." entry in "Cache storage" in the Application panel of Chrome DevTools.
9. Close DevTools
10. Confirm the vast majority of genes in the IL18 signaling pathway diagram have at least some color filled in.  Compare to video for reference of expected colors.

This satisfies SCP-5710, a Developer's Choice Days project.
